GetStream/stream-chat-android
# [CAS-1305] Feature - Ktlint and unit tests for affected modules

[CAS-1305](https://stream-io.atlassian.net/browse/CAS-1305)

### 🎯 Goal

Run unit tests and ktlint only for affected modules. 

### 🛠 Implementation details

This uses [dag-command](https://github.com/leandroBorgesFerreira/dag-command) to create the list of affected modules by a PR and then it uses [android-cli](https://github.com/GetStream/android-cli) to create a command for ktlint and unit tests based on the modules that were affected by a PR. 

**A module doesn't have to be directly changed to be affected by a PR, if it depends on something that was changed, it gets affected.**

### 🧪 Testing

Set this branch it to be the `defaultBranch`, and create a small change in one module. Try to run the unit tests only for the affected modules. You can follow this tutorial:

1. Set this branch in the configuration of dag-command:

```
dagCommand {
    filter = "all"
    defaultBranch = "origin/dagTest"
    outputType = "json"
    printModulesInfo = true
}
```
You can also just add a env var `export GITHUB_BASE_REF="develop"` in your machine too. 

2. Change some file in the project.
3. run `./gradlew dag-command`
4. Check that the file `affected-modules.json` represents only the affected modules by your change. 
5. Run `./gradlew selected-test-debug` and check that the command generated look as expected. You should also run it to confirm that it is correct. 

### ☑️ Checklist


### 🎉 GIF

![giphy](https://user-images.githubusercontent.com/10619102/133156112-c20e9d14-71bb-4d48-a89a-910acfa28145.gif)
